### PR TITLE
feat(metrics_extraction): Return isMetricsExtractedData in event-stats API

### PR DESCRIPTION
**Pull request by armenzg**
When using on-demand metrics, we should return `isMetricsExtractedData: true` in the `events-stat` API.

This is a follow-up to #60047 since we wanted to at least merge the events API portion of it on that PR.

### CODE REVIEW

File: src/sentry/api/serializers/snuba.py [hunk 1]
```diff
@@ -262,6 +262,7 @@ def serialize(
         elif "order" in result.data:
             res["order"] = result.data["order"]
         res["isMetricsData"] = result.data.get("isMetricsData", False)
+        res["isMetricsExtractedData"] = result.data.get("isMetricsExtractedData", False)
```
Comment (armenzg) [inline src/sentry/api/serializers/snuba.py:265]: This value is added in the next file (run_query from [timeseries_query](https://github.com/getsentry/sentry/blob/d71dab6cab04c54d9e18ccaec60874ee34ad2986/src/sentry/snuba/metrics_performance.py#L189) in `src/sentry/snuba/metrics_performance.py`).

File: tests/snuba/api/endpoints/test_organization_events_mep.py [hunk 2]
```diff
@@ -2602,12 +2602,7 @@ def do_request(self, query):
             self.viewname,
             kwargs={"organization_slug": self.organization.slug},
         )
-        with self.feature(
-            {
-                "organizations:on-demand-metrics-extraction": True,  # XXX: We have coupling
-                "organizations:on-demand-metrics-extraction-widgets": True,
-            }
-        ):
+        with self.feature({"organizations:on-demand-metrics-extraction": True}):
```
Comment (armenzg) [inline tests/snuba/api/endpoints/test_organization_events_mep.py:2605]: This is an unrelated change to the PR. This may be included in #60176 or not.

File: tests/snuba/api/endpoints/test_organization_events_stats_mep.py [hunk 3]
```diff
@@ -1150,3 +1153,36 @@ def test_top_events_with_transaction_on_demand_and_no_environment(self):
...
+        return meta
+
+    def test_is_metrics_extracted_data_is_included(self):
+        self._test_is_metrics_extracted_data(
+            {
+                "dataset": "metricsEnhanced",
+                "query": "transaction.duration:>=91",
+                "useOnDemandMetrics": "true",
+                "yAxis": "count()",
+            },
```
Comment (armenzg) [inline tests/snuba/api/endpoints/test_organization_events_stats_mep.py:1185]: Almost identical params to the `events` endpoint except this one does not require `field`. See test:
https://github.com/getsentry/sentry/blob/8eba43d1c1d3288382b7de1d1c63273ce75410f2/tests/snuba/api/endpoints/test_organization_events_mep.py#L2634-L2640

File: src/sentry/api/bases/organization_events.py [hunk 4]
```diff
@@ -525,6 +525,12 @@ def get_event_stats_data(
                             zerofill_results=zerofill_results,
                             dataset=dataset,
                         )
+                        # XXX: Is this done for isMetricsData somewhere else?
+                        for c in query_columns:
+                            # At least one of the columns has required extracted data
+                            if results[key][c].get("meta", {}).get("isMetricsExtractedData"):
```
Comment (armenzg) [inline src/sentry/api/bases/organization_events.py:531]: This information comes from the lines I added below:
https://github.com/getsentry/sentry/blob/b154b24ad2235d92dfec3e7294d2724dcf08d2fe/src/sentry/snuba/metrics_performance.py#L449-L451

File: src/sentry/api/bases/organization_events.py [hunk 5]
```diff
@@ -525,6 +525,12 @@ def get_event_stats_data(
                             zerofill_results=zerofill_results,
                             dataset=dataset,
                         )
+                        # XXX: Is this done for isMetricsData somewhere else?
```
[the remark this answers is not in the document]
Comment (k-fish) [reply src/sentry/api/bases/organization_events.py:528]: `isMetricsData` doesn't show up in group-by on tables except in the root meta 👍

File: tests/snuba/api/endpoints/test_organization_events_stats_mep.py [hunk 6]
```diff
@@ -350,6 +359,7 @@ def test_sum_transaction_duration(self):
         )
         assert response.status_code == 200, response.content
         assert response.data["isMetricsData"]
+        assert response.data["isMetricsExtracteData"] is False
```
Comment (k-fish) [inline tests/snuba/api/endpoints/test_organization_events_stats_mep.py:362]:
```suggestion
        assert response.data["isMetricsExtractedData"] is False
```

File: tests/snuba/api/endpoints/test_organization_events_stats_mep.py [hunk 7]
```diff
@@ -383,6 +393,7 @@ def test_sum_transaction_duration_with_comparison(self):
...
```
Comment (k-fish) [inline tests/snuba/api/endpoints/test_organization_events_stats_mep.py:396]:
```suggestion
        assert response.data["isMetricsExtractedData"] is False
```

File: tests/snuba/api/endpoints/test_organization_events_stats_mep.py [hunk 8]
```diff
@@ -431,6 +442,7 @@ def test_custom_measurement(self):
...
```
Comment (k-fish) [inline tests/snuba/api/endpoints/test_organization_events_stats_mep.py:445]:
```suggestion
        assert response.data["isMetricsExtractedData"] is False
```

File: tests/snuba/api/endpoints/test_organization_events_stats_mep.py [hunk 9]
```diff
@@ -460,6 +472,7 @@ def test_does_not_fallback_if_custom_metric_is_out_of_request_time_range(self):
         meta = response.data["meta"]
         assert response.status_code == 200, response.content
         assert response.data["isMetricsData"]
+        assert response.data["isMetricsExtracteData"] is False
```
Comment (k-fish) [inline tests/snuba/api/endpoints/test_organization_events_stats_mep.py:475]: ditto

File: tests/snuba/api/endpoints/test_organization_events_mep.py [hunk 10]
```diff
@@ -2642,19 +2642,6 @@ def test_is_metrics_extracted_data_is_included(self):
             dataset="metricsEnhanced",
         )
 
-    def test_is_metrics_extracted_data_is_excluded(self):
```
Comment (armenzg) [inline tests/snuba/api/endpoints/test_organization_events_mep.py:2645]: This test does nothing useful. I will need to write a new one on a different PR.

File: src/sentry/snuba/metrics_performance.py [hunk 11]
```diff
@@ -243,10 +243,13 @@ def run_metrics_query(inner_params: Dict[str, Any]):
             )
             sentry_sdk.set_tag("performance.dataset", "metrics")
             result["meta"]["isMetricsData"] = True
+            sentry_sdk.set_tag("on_demand.is_extracted", metrics_query.use_on_demand)
```
Comment (k-fish) [inline src/sentry/snuba/metrics_performance.py:246]: 👍

**Review by k-fish — APPROVED** [review]
Looks like there is still some errors in the tests, but otherwise code and test coverage looks good, if we're missing something we can add it later 👍

File: src/sentry/api/bases/organization_events.py [hunk 12]
```diff
@@ -525,6 +525,9 @@ def get_event_stats_data(
                             zerofill_results=zerofill_results,
                             dataset=dataset,
                         )
+                        results[key]["isMetricsExtractedData"] = self._query_if_extracted_data(
```
Comment (armenzg) [inline src/sentry/api/bases/organization_events.py]: I've moved the logic into a function with a try/except block to protect against side effects.

File: src/sentry/api/bases/organization_events.py [hunk 13]
```diff
@@ -525,6 +525,10 @@ def get_event_stats_data(
                             zerofill_results=zerofill_results,
                             dataset=dataset,
                         )
+                        if request.query_params.get("useOnDemandMetrics") == "true":
```
Comment (armenzg) [inline src/sentry/api/bases/organization_events.py:528]: If top n events are called but we are not using on-demand, there's no reason to even try the code.

File: src/sentry/api/bases/organization_events.py [hunk 14]
```diff
@@ -525,6 +525,10 @@ def get_event_stats_data(
...
+                            results[key]["isMetricsExtractedData"] = self._query_if_extracted_data(
```
Comment (armenzg) [inline src/sentry/api/bases/organization_events.py:529]: Moved the logic within a function and make it safe within a try/except block.

File: src/sentry/api/bases/organization_events.py [hunk 15]
```diff
@@ -579,6 +583,21 @@ def get_event_stats_data(
...
+    ) -> bool:
+        ret_value = False
+        try:
+            for c in query_columns:
+                # At least one of the columns has required extracted data
+                if results[key][c].get("meta", {}).get("isMetricsExtractedData"):
+                    ret_value = True
+                    break
+        except Exception as error:
+            sentry_sdk.capture_exception(error)
```
Comment (armenzg) [inline src/sentry/api/bases/organization_events.py:597]: We would be able to find any unknown issues.

**Review by armenzg — COMMENTED** [review]
@k-fish I have [removed these other test checks](https://github.com/getsentry/sentry/pull/60183/commits/ea5d0071a261c5e8d5eaf886b92422e72b436971) since it would require work for each to determine how to handle them.
